lib/ruby_lsp/listeners/discover_tests.rb

```diff
@@ -0,0 +1,83 @@
+# typed: strict
+# frozen_string_literal: true
+
+module RubyLsp
+  module Listeners
+    class DiscoverTests
+      extend T::Helpers
+      abstract!
+
+      include Requests::Support::Common
+
+      DYNAMIC_REFERENCE_MARKER = "<dynamic_reference>"
+
+      #: (ResponseBuilders::TestCollection response_builder, GlobalState global_state, Prism::Dispatcher dispatcher, URI::Generic uri) -> void
+      def initialize(response_builder, global_state, dispatcher, uri)
+        @response_builder = response_builder
+        @uri = uri
+        @index = T.let(global_state.index, RubyIndexer::Index)
+        @visibility_stack = T.let([:public], T::Array[Symbol])
+        @nesting = T.let([], T::Array[String])
+
+        @fully_qualified_name = T.let("", String)
+        @attached_ancestors = T.let([], T::Array[String])
+      end
+
+      #: (Prism::ClassNode node) -> void
+      def on_class_node_enter(node) # rubocop:disable RubyLsp/UseRegisterWithHandlerMethod
```

Comment:
(the need to disable RoboCop shows that this cop is far from ideal. Hopefully we can merge https://github.com/Shopify/ruby-lsp/pull/3291 and eliminate it).